cdap-watchdog/src/main/java/co/cask/cdap/logging/gateway/handlers/AbstractChunkedLogProducer.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.logging.gateway.handlers;
+
+import co.cask.cdap.api.dataset.lib.CloseableIterator;
+import co.cask.cdap.logging.read.LogEvent;
+import co.cask.http.BodyProducer;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Multimap;
+import org.jboss.netty.buffer.ChannelBuffer;
+import org.jboss.netty.buffer.ChannelBuffers;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.nio.ByteBuffer;
+import java.nio.CharBuffer;
+import java.nio.charset.Charset;
+import java.nio.charset.CharsetEncoder;
+import java.nio.charset.CoderResult;
+import javax.annotation.Nullable;
+
+/**
+ * LogReader BodyProducer class that delegates to subclasses for how to encode log events.
+ */
+public abstract class AbstractChunkedLogProducer extends BodyProducer {
+  private static final Logger LOG = LoggerFactory.getLogger(AbstractChunkedLogProducer.class);
+
+  private final ByteBuffer chunkBuffer = ByteBuffer.allocate(8 * 1024);
+  private final CharsetEncoder charsetEncoder = Charset.forName("UTF-8").newEncoder();
+
+  private final CloseableIterator<LogEvent> logEventIter;
+  private CharBuffer charBuffer = CharBuffer.allocate(0);
+
+  private boolean doneSending = false;
+
+  AbstractChunkedLogProducer(CloseableIterator<LogEvent> logEventIter) {
+    this.logEventIter = logEventIter;
+  }
+
+  /**
+   * Return {@link Multimap} of HTTP response headers
+   */
+  protected abstract Multimap<String, String> getResponseHeaders();
+
+  protected abstract CharBuffer transformEvent(LogEvent event);
```

Comment:
Why use `CharBuffer` as the contract? I think it's better be `ChannelBuffer`. We don't need the intermediate conversion between different types of buffers.